media-gfx/xpaint: fix some bugs and add new version 3.1.3

ceamac opened the pull request:
`media-libs/libpgf` and `x11-libs/libXaw3dXft` are dependencies of `media-gfx/xpaint`, so I added new versions for them too.

`x11-libs/libXaw3dXft` was required, because `media-gfx/xpaint-3.1.3` would not compile with `x11-libs/libXaw3dXft-1.6.2d`.

About the remaining bugs:
- https://bugs.gentoo.org/735592 this seems to have been fixed, nothing to do for it.
- https://bugs.gentoo.org/382111 I checked, both `media-libs/libpgf-7.19.3` and `media-libs/libpgf-7.21.7` compile fine on ppc.
- https://bugs.gentoo.org/467746 this is old, both versions compile just fine.  I haven't checked exactly what version of autoconf they use.

thesamesam commented on media-gfx/xpaint/files/xpaint-2.10.2-respect-ldflags.patch:
```diff
@@ -0,0 +1,15 @@
+--- a/util/Makefile	2017-03-14 20:54:48.000000000 +0200
```
I'd drop the [timestamps](https://devmanual.gentoo.org/ebuild-writing/misc-files/patches/index.html#clean-patch-howto) but it's not a big deal, more an FYI.

thesamesam commented on media-gfx/xpaint/files/xpaint-2.10.2-respect-ldflags.patch:
```diff
@@ -0,0 +1,15 @@
+ 
+ pdfconcat:
+-	$(CC) $(CFLAGS) -O3 -s -DNDEBUG=1 -DNO_CONFIG=1 -ansi -pedantic -Wunused -Wall -W -Wstrict-prototypes -Wtraditional -Wnested-externs -Winline -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes -Wmissing-declarations pdfconcat.c -o pdfconcat
++	$(CC) $(CFLAGS) $(LDFLAGS) -O3 -s -DNDEBUG=1 -DNO_CONFIG=1 -ansi -pedantic -Wunused -Wall -W -Wstrict-prototypes -Wtraditional -Wnested-externs -Winline -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes -Wmissing-declarations pdfconcat.c -o pdfconcat
```
```suggestion
+	$(CC) $(CFLAGS) $(LDFLAGS) -DNDEBUG=1 -DNO_CONFIG=1 -ansi -pedantic -Wunused -Wall -W -Wstrict-prototypes -Wtraditional -Wnested-externs -Winline -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes -Wmissing-declarations pdfconcat.c -o pdfconcat
```

We don't want to force -O3 or to strip.

ceamac replied:
`-O3` and `-s` were already stripped with *sed* from the ebuild.  But yes, I'll move them here

thesamesam replied:
Yeah, it's just nicer to do a patch where we can (especially given we're already doing it) as sed can 'fail silently'.

thesamesam commented on media-gfx/xpaint/files/xpaint-2.10.2-respect-ldflags.patch:
```diff
@@ -0,0 +1,15 @@
+ 
+ ppmtops:
+-	$(CC) $(CFLAGS) ppmtops.c -o ppmtops
++	$(CC) $(CFLAGS) $(LDFLAGS) ppmtops.c -o ppmtops
```
```suggestion
+	$(CC) $(CFLAGS) $(LDFLAGS) ppmtops.c -o ppmtops
```

In theory, we need $(CPPFLAGS) both here and above.

thesamesam commented on media-gfx/xpaint/files/xpaint-3.1.3-cflags-ldflags-strip.patch:
```diff
@@ -0,0 +1,38 @@
+--- a/vxp2ps/Makefile	2021-01-19 10:35:54.000000000 +0200
```
kill timestamps

thesamesam commented on media-gfx/xpaint/files/xpaint-3.1.3-cflags-ldflags-strip.patch:
```diff
@@ -0,0 +1,38 @@
+ 
+ vxp2ps: vxp2ps.c
+-	gcc vxp2ps.c -o vxp2ps -lm $(CFLAGSEXTRA)
++	gcc vxp2ps.c -o vxp2ps -lm $(CFLAGS) $(CFLAGSEXTRA) $(LDFLAGS)
```
```suggestion
+	$(CC) vxp2ps.c -o vxp2ps -lm $(CFLAGS) $(CFLAGSEXTRA) $(LDFLAGS)
```
repeat everywhere

thesamesam commented on media-gfx/xpaint/files/xpaint-3.1.3-cflags-ldflags-strip.patch:
```diff
@@ -0,0 +1,38 @@
+ 
+ vxp2ps: vxp2ps.c
+-	gcc vxp2ps.c -o vxp2ps -lm $(CFLAGSEXTRA)
++	gcc vxp2ps.c -o vxp2ps -lm $(CFLAGS) $(CFLAGSEXTRA) $(LDFLAGS)
```
CPPFLAGS after $(CC) and possibly kill CFLAGSEXTRA (what is it?)

ceamac replied:
`CFLAGSEXTRA` is a dependency to `libdvipgm` we don't have for now, so it expands to nothing.  It can be added later if needed.

thesamesam commented on media-gfx/xpaint/files/xpaint-3.1.3-cflags-ldflags-strip.patch:
```diff
@@ -0,0 +1,38 @@
+ 
+ pdfconcat:
+-	$(CC) $(CFLAGS) -O3 -s -DNDEBUG=1 -DNO_CONFIG=1 -ansi -pedantic -Wunused -Wall -W -Wnested-externs -Winline -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes -Wmissing-declarations pdfconcat.c -o pdfconcat
++	$(CC) $(CFLAGS) $(LDFLAGS) -O3 -s -DNDEBUG=1 -DNO_CONFIG=1 -ansi -pedantic -Wunused -Wall -W -Wnested-externs -Winline -Wpointer-arith -Wbad-function-cast -Wcast-qual -Wmissing-prototypes -Wmissing-declarations pdfconcat.c -o pdfconcat
```
see earlier comments

thesamesam commented on media-gfx/xpaint/files/xpaint-libtool-clang.patch:
```diff
@@ -0,0 +1,11 @@
+--- a/xpaintrw/Makefile.am	2016-12-10 12:18:15.000000000 +0200
```
Remember to have libtool in BDEPEND as it's calling the host's one.

thesamesam commented on media-gfx/xpaint/xpaint-2.10.2-r1.ebuild:
```diff
@@ -42,6 +42,11 @@ BDEPEND="
 "
 
+PATCHES=(
+	"${FILESDIR}"/${PN}-libtool-clang.patch
```
Put the version you based this on in the filename

ceamac replied:
I had it but I removed it because I am using the same patch bot for *2.10.2-r1* and *3.1.3*.  Is it ok to patch *3.1.3* with a patch named *2.10.2*?

thesamesam replied:
Yeah, that's no problem. I name it after the first version in tree it applies to/the one I based it on.

thesamesam commented on media-gfx/xpaint/xpaint-2.10.2-r1.ebuild:
```diff
@@ -42,6 +42,11 @@ BDEPEND="
+
 src_prepare() {
 	default
 	sed -i -e 's/-O3 -s//g' util/Makefile || die
```
We can just fix the Makefile itself given we're already patching it.

thesamesam commented on media-gfx/xpaint/xpaint-3.1.3.ebuild:
```diff
@@ -0,0 +1,97 @@
+
+EAPI=7
+
+inherit desktop toolchain-funcs autotools xdg-utils
```
Sort alphabetically

thesamesam commented on media-gfx/xpaint/xpaint-3.1.3.ebuild:
```diff
@@ -0,0 +1,97 @@
+	)
+"
+DEPEND="${RDEPEND}"
+BDEPEND="
```
libtool?

thesamesam commented on media-gfx/xpaint/xpaint-3.1.3.ebuild:
```diff
@@ -0,0 +1,97 @@
+
+src_prepare() {
+	default
+	sed -i -e 's/-O3 -s//g' util/Makefile || die
```
We could just patch it out as we're already there

thesamesam commented on media-gfx/xpaint/xpaint-3.1.3.ebuild:
```diff
@@ -0,0 +1,97 @@
+	emake clean
+	emake -C util clean
+
+	# parallel make still fails sometimes
```
We say that, but no -j1...?

ceamac replied:
I think this this comment is about the next 2 lines, probably they were breaking parallel make so they are done before the main emake.
```
    emake substads
    emake xpaint.1
```

I compiled this many times using `-j12`, both with gcc and clang, and it seems fine.  I wouldn't add `-j1` unless someone reports a bug.

thesamesam replied:
Oh, okay, I see.

thesamesam commented on media-libs/libpgf/libpgf-7.21.7.ebuild:
```diff
@@ -0,0 +1,41 @@
+	# configure.ac has wrong version number
+	sed -i "s/7.15.32/${PV}/g" configure.ac || die
+
+	# many files, including configure.ac and Makefile.am, are saved in dos format, causing errors in autotools
```
edos2unix.eclass?

thesamesam commented:
Could you also add the bug references to the top of each of the patches and a brief description of what they do/are for?

ceamac commented:
I think I resolved all, and I added a short description before each patch.  Please review

juippis commented on x11-libs/libXaw3dXft/libXaw3dXft-1.6.2h.ebuild:
```diff
@@ -0,0 +1,36 @@
+
+src_configure() {
+	local XORG_CONFIGURE_OPTIONS=(
+		$(use_enable unicode internationalization)
```
This should probably be forced on already.

ceamac replied:
`netpbm` was indeed a new dependency, but I don't understand the problem with `epstopdf: command not found`

Neither *xpaint* nor *libpgf* require `epstopdf`.  I even tried removing the executable from my system and it still compiles fine.  I see `doxygen` requires `epstopdf`, but it also installs it as a dependency, so I don't understand how to get that error, since **USE=doc** requires `doxygen`.

ceamac replied:
I checked.  `--enable-internationalization` is the default.  With **USE=-unicode**, that will pass `--disable-internationalization` to configure, which is probably what we want.

juippis reviewed (commented):
Didn't test with clang, but with gcc xpaint-3.1.3 always fails for me:
```
readWritePNM.c:20:10: fatal error: netpbm/pam.h: No such file or directory
   20 | #include <netpbm/pam.h>
      |          ^~~~~~~~~~~~~~
...
make[2]: Leaving directory '/var/tmp/portage/media-gfx/xpaint-3.1.3/work/xpaint-3.1.3/xpaintrw'
make[1]: *** [Makefile:656: all-recursive] Error 1
make[1]: Leaving directory '/var/tmp/portage/media-gfx/xpaint-3.1.3/work/xpaint-3.1.3'
make: *** [Makefile:448: all] Error 2
 * ERROR: media-gfx/xpaint-3.1.3::gentoo failed (compile phase):
 *   emake failed
```
```
 * Package:    media-libs/libpgf-7.21.7
 * Repository: gentoo
 * Maintainer: ceamac.paragon@gmail.com proxy-maint@gentoo.org
 * USE:        abi_x86_64 amd64 doc elibc_glibc kernel_linux userland_GNU
...
 * QA Notice: command not found:
 * 
 * 	sh: line 1: epstopdf: command not found
 * 	sh: line 1: epstopdf: command not found
 * 	sh: line 1: epstopdf: command not found
 * 	sh: line 1: epstopdf: command not found
 * 	sh: line 1: epstopdf: command not found
```

ceamac commented:
I finished building a gentoo prefix, I'll start using it to check ebuilds too, as it has far fewer dependencies already installed.

I'm marking this as draft until I fix it.  Thanks!

ceamac commented:
I fixed the compile error, was a missing dependency.

While trying to test it inside a gentoo prefix, I got lots of errors because xpaint has many hardcoded paths.  I fixed them and rearranged the patches a little.  This is a new commit at the end, it can be skipped if it looks too wonky.

Please review.

thesamesam commented:
Given @juippis started, I'll let him finish unless he'd like me to

thesamesam commented:
Actually, on reflection, he's been excellent enough to do mine and I'm worrying too much about interfering ;)

I find it helpful and hopefully he does too. I'll do it now!

thesamesam commented:
Great work here, by the way!

ceamac commented:
Thank you!

juippis commented:
@thesamesam by no means did I mean to insult you, just going through unread e-mails and spotted an unmerged PR. Trying to help in case the original reviewer doesn't have time, or forgot to merge... you know the drill.

And thanks for re-checking, haven't been at my PC in 3 days ;)
